cmd/notation/login.go

```diff
@@ -70,61 +73,69 @@ func runLogin(ctx context.Context, opts *loginOpts) error {
 	// input username and password by prompt
 	reader := bufio.NewReader(os.Stdin)
 	var err error
-	if opts.Username == "" {
-		opts.Username, err = readUsernameFromPrompt(reader)
-		if err != nil {
-			return err
-		}
-	}
-
 	if opts.Password == "" {
-		opts.Password, err = readPasswordFromPrompt(reader)
+		var isToken bool
+		if opts.Username == "" {
+			opts.Username, err = readUsernameFromPrompt(reader)
+			if err != nil {
+				return err
+			}
+			if opts.Username == "" {
+				// the username is empty, the password is used as a token
+				isToken = true
+			}
+		}
+		opts.Password, err = readPasswordFromPrompt(reader, isToken)
 		if err != nil {
 			return err
 		}
+		if opts.Password == "" {
+			if isToken {
+				return errors.New("token required")
+			}
+			return errors.New("password required")
+		}
 	}
+	cred := opts.Credential()
 
-	if err := validateAuthConfig(ctx, opts, serverAddress); err != nil {
-		return err
+	credsStore, err := auth.NewCredentialsStore()
+	if err != nil {
+		return fmt.Errorf("failed to get credentials store: %v", err)
 	}
-
-	nativeStore, err := auth.GetCredentialsStore(ctx, serverAddress)
+	registry, err := getRegistryLoginClient(ctx, &opts.SecureFlagOpts, serverAddress)
 	if err != nil {
-		return fmt.Errorf("could not get the credentials store: %v", err)
+		return fmt.Errorf("failed to get registry client: %v", err)
 	}
+	if err := credentials.Login(ctx, credsStore, registry, cred); err != nil {
+		registryName := registry.Reference.Registry
+		if !errors.Is(err, credentials.ErrPlaintextPutDisabled) {
+			return fmt.Errorf("failed to log in to %s: %v", registryName, err)
+		}
 
-	// init creds
-	creds := newCredentialFromInput(
-		opts.Username,
-		opts.Password,
-	)
-	if err = nativeStore.Store(serverAddress, creds); err != nil {
-		return fmt.Errorf("failed to store credentials: %v", err)
+		// ErrPlaintextPutDisabled returned by Login() indicates that the
+		// credential is validated but is not saved because there is no native
+		// credentials store available
+		if savedCred, err := credsStore.Get(ctx, registryName); err != nil || savedCred != cred {
+			if err != nil {
+				// if we fail to get the saved credential, log a warning
+				// but do not throw the GET error, as the error could be
+				// confusing to users
+				logger := log.GetLogger(ctx)
+				logger.Warnf("Failed to get the existing credentials for %s: %v", registryName, err)
+			}
+			return fmt.Errorf("failed to log in to %s: the credential could not be saved because a credentials store is required to securely store the password. See %s",
+				registryName, urlDocHowToAuthenticate)
+		}
+
+		// the credential already exists somewhere, ignore the saving error
+		fmt.Fprintf(os.Stderr, "Warning: The credentials store is not set up. It is recommended to configure the credentials store to securely store your credentials. See %s.\n", urlDocHowToAuthenticate)
+		fmt.Println("Authenticated with existing credentials")
```

Comment:
Question: When the printouts are line 131 + line 132, the user might be curious about where exactly the existing credentials are saved? Do we want to provide more details here?

Reply:
From the notation standpoint, we don't know where exactly the existing credential is... Will this information be useful in any scenarios?

Reply:
> From the notation standpoint, we don't know where exactly the existing credential is... Will this information be useful in any scenarios?

Yeah, because the printouts claim that the `login` succeeded without a credential store. As a user, it's nature to ask, then where are my credentials saved?

Reply:
I think the user should know better than notation as they have logged in using other tools by themselves. 😂